Expert dataset: load from memory

AdamGleave opened the pull request:
Fixes #281. I add a keyword argument `traj_data` to `ExpertDataset` that allows for directly passing in the in-memory dataset, falling back to loading from `expert_path` if not specified. A cleaner alternative would be to delete `expert_path` (it just requires one additional line of code for users), but I'm concerned this would break backwards compatibility.

araffin commented on stable_baselines/gail/dataset/dataset.py:
```diff
@@ -29,10 +30,12 @@ class ExpertDataset(object):
+        if traj_data is not None and expert_path is not None:
+            raise ValueError("Cannot specify both 'traj_data' and 'expert_path'")
+        if traj_data is None:
+            traj_data = np.load(expert_path)
```
you should also check that expert_path is not None, no?

AdamGleave replied:
My logic here was that we weren't previously checking for this, and `np.load` would generate an error. However on reflection I agree now I've made `expert_path` a keyword rather than positional argument I should check for this; added.

araffin commented:
I would also add an option in the trajectory generator to return the dataset, so you don't have to load it from the disk.

AdamGleave commented:
Thanks, I've updated the trajectory generator. In the case of image-based observations, the code by default saves them as PNGs to files, which the comments indicate more efficient (presumably because of compression). I've modified it to save them as raw images if no save path is specified, but warn about this.

araffin commented on stable_baselines/gail/dataset/record_expert.py:
```diff
@@ -53,6 +56,10 @@ def generate_expert_traj(model, save_path, env=None, n_timesteps=0,
     record_images = len(obs_space.shape) == 3 and obs_space.shape[-1] in [1, 3, 4] \
                     and obs_space.dtype == np.uint8
+    if record_images and save_path is None:
+        warnings.warn("Observations are images but no save path, so will save in numpy archive; "
```
```suggestion
        warnings.warn("Observations are images but no save path was specified, so will save in numpy archive; "
```

araffin commented:
> In the case of image-based observations, the code by default saves them as PNGs to files, which the comments indicate more efficient (presumably because of compression)

well, default to jpg, png when n_channel=4. It avoids loading everything in the memory, which can easily result in eating up the whole RAM.

AdamGleave commented:
Updated changelog and added tests.

What documentation do you think should be updated? I don't see much GAIL specific in docs/ and I think I've already updated the relevant docstrings.

araffin commented:
I was thinking about the [pretraining doc](https://stable-baselines.readthedocs.io/en/master/guide/pretrain.html) but in fact that is not needed.

Btw, I just found a typo in the [code example](https://stable-baselines.readthedocs.io/en/master/guide/pretrain.html#pre-train-a-model-using-behavior-cloning) could you also fix that? (there is a missing parentheses)

Note: the error from codacy is only because of a too long name I think

araffin commented on tests/test_gail.py:
```diff
@@ -12,12 +13,17 @@
+                                        ('CartPole-v1', EXPERT_PATH_DISCRETE, False)])
 def test_gail(expert_env):
-    env_id, expert_path = expert_env
+    env_id, expert_path, we_load = expert_env
```
I would go for `load_from_memory` to be more explicit ;)

AdamGleave commented:
Fixed. 

I'm not sure why the Travis build was failing; it worked fine on my local Docker image. Suspect some test non-determinism as one of the Travis builds passed; patched `test_pretrain` to avoid a possible race condition.

araffin commented:
>I'm not sure why the Travis build was failing;

with `tests/test_gail.py::test_pretrain_images` I would suspect a memory error...

araffin commented on stable_baselines/gail/dataset/record_expert.py:
```diff
@@ -21,13 +21,16 @@ def generate_expert_traj(model, save_path, env=None, n_timesteps=0,
     :param n_timesteps: (int) Number of training timesteps
     :param n_episodes: (int) Number of trajectories (episodes) to record
     :param image_folder: (str) When using images, folder that will be used to record images.
+    :return the generated expert trajectories.
```
```suggestion
    :return: (dict) the generated expert trajectories.
```

AdamGleave commented:
@araffin that 4-character change made GitHub dismiss your review, so I need you to approve it again :(

araffin commented:
>dismiss your review, so I need you to approve it again

Yes, I couldn't commit it directly (maybe you did not check "allow edit from maintainers" ?)
